### ISSUE TITLE
Failure getting the list of custom fields

### ISSUE THREAD

**Issue by jabatrox** [guilhemmarchand/TA-jira-service-desk-simple-addon#16]
When going to Get JIRA info/Get fields description, instead of having displayed the list of fields as showed in the documentation here:
![image](https://user-images.githubusercontent.com/3520254/84819396-35690480-b018-11ea-8b22-3197fb0a18ee.png)

This is what is displayed:
![image](https://user-images.githubusercontent.com/3520254/84819225-f89d0d80-b017-11ea-862b-c4cd81cda470.png)

I am not sure if I am doing anything wrong, but I remember I tried it a couple of weeks ago, before the last version updates, and it was working, so I believe it might be some problem parsing the custom fields as a result of the bugfix that lead to version 1.10, but it is just a guess.

**Comment by guilhemmarchand** [comment]
Hi @jabatrox 

This actually is normal, the report is exposing the result as events (json):

![image](https://user-images.githubusercontent.com/7503059/84823329-110c2880-b016-11ea-8ee9-91a047cdd8f4.png)

If you click on Edit then open in search on top right:

![image](https://user-images.githubusercontent.com/7503059/84823428-37ca5f00-b016-11ea-9c02-0fa1ad3808a7.png)

Then you land here:

![image](https://user-images.githubusercontent.com/7503059/84823481-503a7980-b016-11ea-9d57-a8cb7b2c3a22.png)

And if you click on All fields on the left part of the screen, you will see the exact same picture:

![image](https://user-images.githubusercontent.com/7503059/84823541-6ba58480-b016-11ea-8079-601defc6b088.png)

Let me know if this answers your question and feel free to close the issue if that's all clear.

Guilhem

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

I already managed to reach that point, but for some reason I don't see all these fields, only these ones:
![image](https://user-images.githubusercontent.com/3520254/84831107-09a34a00-b02b-11ea-8bd9-2312f74f7fe9.png)

And, if I click on All Fields:
![image](https://user-images.githubusercontent.com/3520254/84831135-1758cf80-b02b-11ea-96b4-1fbc8da1a217.png)

**Comment by guilhemmarchand** [comment]
I see @jabatrox 

When you develop one json, you get values right ?

![image](https://user-images.githubusercontent.com/7503059/84837344-c8616900-b02f-11ea-96bc-bc0897f5a9e3.png)

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

No, I don't get any values when I expand the JSON:
![image](https://user-images.githubusercontent.com/3520254/84922497-39ebf680-b0c6-11ea-97af-a3bae1169b9c.png)

I have no idea why, it worked a couple of week ago, and we have not made any change to JSD that could lead to this issue.

**Comment by guilhemmarchand** [comment]
Ok I will double check @jabatrox 
Are the others reports providing results as expected ? 

Get projects / Get issues types / Get issues priorities

**Comment by jabatrox** [comment]
Thank you @guilhemmarchand 

Yes, all other reports give the expected results: projects their keys show up, as well as issue types and priorities. Fields description is the only one failing.

**Comment by hamidbg86** [comment]
@jabatrox  and @guilhemmarchand 

I have same issue, custom fields are not being extracted. other reports are working fine. 

Thanks!

**Comment by guilhemmarchand** [comment]
@hamidbg86 
@jabatrox 

Fair enough, the Python code underneath the custom command does the following REST call to JIRA:

`
curl -u <user>:'<password>' 'https://<my_jira_instance>/rest/api/2/issue/createmeta?projectKeys=<my_project>&issuetypeNames=Task&expand=projects.issuetypes.fields' | jq
`

How does this looks like on your end if you are able to run the same curl command on a Linux box ?

It's a bit odd I couldn't reproduce yet:

![image](https://user-images.githubusercontent.com/7503059/85310751-4c8e7300-b4ac-11ea-9de5-e621bfa23485.png)

**Comment by hamidbg86** [comment]
@guilhemmarchand    I get same results like shown in your screenshot.

**Comment by guilhemmarchand** [comment]
Hum ok I a bit confused now ;-)

My screen shot is the result expected, say I have a custom field named "test_multi_select", if I open the report, then open in search and add the name of the custom field:

![image](https://user-images.githubusercontent.com/7503059/85321975-7b611500-b4bd-11ea-8de6-60c874c79211.png)

Then I can see the raw json response, I can see all the fields extracted etc

![image](https://user-images.githubusercontent.com/7503059/85321892-566ca200-b4bd-11ea-9c4f-4981f819a220.png)

Don't we see the same thing then ?

**Comment by hamidbg86** [comment]
@guilhemmarchand I did the same request on postman and I was able to find a custom field. But the issue is when we run that in splunk it doesn't return same information. not sure why. 

Ty!

**Comment by jabatrox** [comment]
> @hamidbg86
> @jabatrox
> 
> Fair enough, the Python code underneath the custom command does the following REST call to JIRA:
> 
> `curl -u <user>:'<password>' 'https://<my_jira_instance>/rest/api/2/issue/createmeta?projectKeys=<my_project>&issuetypeNames=Task&expand=projects.issuetypes.fields' | jq`
> 
> How does this looks like on your end if you are able to run the same curl command on a Linux box ?
> 
> It's a bit odd I couldn't reproduce yet:
> 
> ![image](https://user-images.githubusercontent.com/7503059/85310751-4c8e7300-b4ac-11ea-9de5-e621bfa23485.png)

@guilhemmarchand Ok I think the issue in my case is that my issue type is not "task", since we use custom issue types. Therefore, when I replace "Task" in the link with an actually existing issue type, I get back a JSON as I would see if I accessed the REST API, including all fields (custom or not) and their details, for that specific issue type. Additionally, if I remove the "&issuetypeNames=Task" completely, I get all fields for all issue types associated with the selected project.

The curl command would therefore be:
`curl -u <user>:'<password>' 'https://<my_jira_instance>/rest/api/2/issue/createmeta?projectKeys=<my_project>&expand=projects.issuetypes.fields' | jq`

Or maybe extracting the issue types, and then making a curl call per issue type... but in the end, it is basically the same.

**Comment by guilhemmarchand** [comment]
Thanks, will be reviewing these carefully.

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

Is there any update on this issue? Were you able to fix it, base on my last post?

Thank you

**Comment by guilhemmarchand** [comment]
Hi @jabatrox 

Thanks, you gave me the right information to fix it.
Sorry for the delay, I will be fixing this shortly in a new release, within a few days max.

Guilhem

**Comment by guilhemmarchand** [comment]
@jabatrox 

Do you have a test environment ? Or just a Splunk dev instance you can use to verify ?

**If so, the branch version_1013 contains the fix for your issue:**

https://github.com/guilhemmarchand/TA-jira-service-desk-simple-addon/tree/version_1013

If could give it a test and confirm before we conclude this, this would be great

Guilhem

**Comment by guilhemmarchand** [comment]
@jabatrox 
Version 1.0.14 has been published and addressed it as mentioned earlier. (Version 1.0.13 was not published, it's a bad number and always bad releases!)

**Comment by hamidbg86** [comment]
lol yeah 13 is good to skip it hahah

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

Sorry for the late reply, I didn't see your message at first, and then I was on holidays. I just tested it today, and I still don't get the fields :(. This is what I get:

![image](https://user-images.githubusercontent.com/3520254/90101468-0680bd80-dd3f-11ea-956e-2eb2c245a407.png)

I have tried expanding the trees, and I still don't get the results, as I do for "get projects", "get issue types", and "get priorities". I have tried again with the curl command:
`curl -u <user>:'<password>' 'https://<my_jira_instance>/rest/api/2/issue/createmeta?projectKeys=<my_project>&expand=projects.issuetypes.fields' | jq`

I saw that, when I replace "<my_project>" after "projectKeys=" with a specific key, I actually get the results for that project:
![image](https://user-images.githubusercontent.com/3520254/90102823-bd7e3880-dd41-11ea-84b6-e48842dc4099.png)
(this is just a small section)

However, if I live it blank, or with a "*", what I get seems to be what I see in Splunk: 
![image](https://user-images.githubusercontent.com/3520254/90102645-595b7480-dd41-11ea-8d27-cba28b780adf.png)

**Comment by jabatrox** [comment]
Ok, the curl command also displays the results if you completely remove the "projectKeys=<my_project>" part:
![image](https://user-images.githubusercontent.com/3520254/90103327-996f2700-dd42-11ea-9a72-04e06e92096a.png)

The curl command would be:
`curl -u <user>:'<password>' 'https://<my_jira_instance>/rest/api/2/issue/createmeta?expand=projects.issuetypes.fields' | jq`

This returns all the fields for all the projects. Maybe it would be useful to separate them by project, and make as many different requests as projects exist. But, at least, the above curl command returns the JSON with all the fields.

**Comment by guilhemmarchand** [comment]
Hi @jabatrox 

Ok thank you very for the feedback, my bad I thought this was going to be the one ;-)
Right, I will double check again based on your replies.

Guilhem

**Comment by guilhemmarchand** [comment]
Hi @jabatrox 

I think the first project "ACME" you are taking as for an example might be the root cause of the issue, checking your screenshot it seems there are fields that can be developed ?

The curl query:
curl -u <user>:'<password>' 'https://<my_jira_instance>/rest/api/2/issue/createmeta?projectKeys=<my_project>&expand=projects.issuetypes.fields' | jq

was an example for the explanation, when the command runs, the <my_project> gets replaced by the key of the project, and the query runs for each project available.

For example, I am testing with a JIRA server fresh instance, using:

`curl -k 'https://localhost:8081/rest/api/2/issue/createmeta?projectKeys=SOC&expand=projects.issuetypes.fields' --user guilhem.marchand:'xxxxxxxx | jq`

The name of the project is SOC, and I can see both in Splunk and as part of the curl results actual data:

![image](https://user-images.githubusercontent.com/7503059/91363651-f5cb4f80-e7f4-11ea-853d-d576ce38aa6a.png)

In Splunk:

![image](https://user-images.githubusercontent.com/7503059/91363710-11365a80-e7f5-11ea-8cb9-53280d5814de.png)

I have tested against JIRA Cloud, JIRA service desk and JIRA server and it seems to render as expected.

Effectively, the project could be removed which would render the fields for all the projects rather than broken per project, but I believe this works as it should.

Can you have a double check based on the explanation above ?

Guilhem

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

It does not work for any of the 3 projects we have. Here is a screenshot of what I get in Splunk:

![image](https://user-images.githubusercontent.com/3520254/91445422-64042680-e876-11ea-815e-ee32517ce7ca.png)

As you may see, none of the fields are expanded for any of the projects.

However, if I run a curl command as:
`curl -u <user>:'<password>' 'https://<my_jira_instance>/rest/api/2/issue/createmeta?projectKeys=<my_project>&expand=projects.issuetypes.fields' | jq`
I know that the <my_project> gets replaced for every project key among the 3 we have. The point is that I actually get a correct answer for every project if I use the curl command, but not in Splunk:

- For the ACME project:
![image](https://user-images.githubusercontent.com/3520254/91446428-c4e02e80-e877-11ea-88b3-cdd95d423192.png)

- For another project (you can see the ID is different):
![image](https://user-images.githubusercontent.com/3520254/91446621-f3f6a000-e877-11ea-8779-e74472a17fb3.png)

It came up to my mind that it may be the fact that we have the SSL certificate validation disabled, and that requires to add a "-k" parameter to the curl command to make it work. So the curl command I actually use is:
`curl -u <user>:'<password>' -k 'https://<my_jira_instance>/rest/api/2/issue/createmeta?projectKeys=<my_project>&expand=projects.issuetypes.fields' | jq`

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

Do you have any update on that? Where you able to check it?

Thank you

**Comment by guilhemmarchand** [comment]
Hi @jabatrox

I am glad you didn't give up ;-)
Yes, kind of, I didn't have the time to do it but I was planning to have a switch in the custom command to remove the per project on demand, and have the two variations of the report out of the box.

Will update here shortly

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

I try to help as much as I can to make it work :)
Ok, I will wait for your updates then. I am not sure to understand what you meant, but I would suggest to double check what I mentioned at the end of my last post, about the "-k" option to disable SSL verification in the curl command, as I suspect it might be the reason why we do not get results in Splunk. I guess the addon does not use a curl command properly speaking, but something similar (Python's "requests" library or any other) which will have its own option to disable SSL verification.

Of course, I might be totally wrong and te reason might be different, so as I said, I will wait for your updates.

Thank you for your work on it :)

**Comment by guilhemmarchand** [comment]
@jabatrox 

And it is much appreciated thank you.

To answer, no the SSL validation is not issue here, in Python the custom command achieves a similar check with no SSL validation:

```
                    jira_fields_response = requests.get(
                        url=str(url) + "/rest/api/2/issue/createmeta?projectKeys=" + project_name
                            + "&expand=projects.issuetypes.fields",
                        auth=(username, password),
                        verify=False
                    )
```

So, I don't remember if you had a testing environment, or a way to test against your JIRA instance:

https://github.com/guilhemmarchand/TA-jira-service-desk-simple-addon/tree/version_1019

There are 2 versions of the report:

![image](https://user-images.githubusercontent.com/7503059/92416978-d23ec800-f157-11ea-9109-5ad59eb14356.png)

The last report uses the following code:

```
            if not url.startswith("https://"):
                jira_fields_response = requests.get(
                    url="https://" + str(url) + "/rest/api/2/issue/createmeta?expand=projects.issuetypes.fields",
                    auth=(username, password),
                    verify=False
                )
            else:
                jira_fields_response = requests.get(
                    url=str(url) + "/rest/api/2/issue/createmeta?expand=projects.issuetypes.fields",
                    auth=(username, password),
                    verify=False
                )
```

As you can see the rest call is achieved without specifying a project, so I assume this should finally work for you based on our previous exchanges.

Let me know

Guilhem

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

Yeah, I know the requests library has a "verify=False" property, that is why I was suggesting it might be that, but if you have already considered that, it is obviously not the issue.

We do have a testing environment, and I was able to test it today. In the "per project" section, I still have the same problem shown in the 1st picture of my previous message (fields not expanding). 

However, in the "all projects" section, I get a "kinda result":
![image](https://user-images.githubusercontent.com/3520254/92627950-616cec80-f2cc-11ea-88fc-d3147cfbb975.png)

As you can see, I can actually see the custom fields by expanding every section, but it is far from being the way of showing it that appears in the docs:
![image](https://user-images.githubusercontent.com/3520254/92628059-8bbeaa00-f2cc-11ea-9d3a-d60e3ff34dc8.png)

In the results I get, the custom fields appear only as part of the JSON structure returned from the API REST, which means I need to expand one of the projects in "projects", one of the custom issue types I have in "issuetypes", and then the "fields" key within it. Then, I get all the fields for that issue type, including JIRA's default ones (such as "assignee"), not just the custom fields.

I honestly have no idea why this is happening. Maybe the Splunk query/command that is run to retrieve and display the fields has to be somehow adapted. I don't know exactly how do you perform that Splunk query over the returned JSON, but maybe I can help you out with it if you tell me which one it is, and how I could test it and modify it "on live" in Splunk. I can't promise to find anything, but I'd give a try at least.

Thanks again for the work on it. We really appreciate :)

**Comment by guilhemmarchand** [comment]
Hi @jabatrox !

Ok, not a problem thanks ;-)

Right, sorry to say that but your answer is kind of confusing, the results I see from the raw search (the first screenshot) is what you would expect, so this is good.

Then the second screenshot which I understand came from the docs is when you click on All fields on the left part of your search window, it's a builtin Splunk view, if that is clear then maybe you had some filters on top of the window preventing you from viewing the fields properly ? In the docs I was explicitly looking for custom fields too.

Guilhem

**Comment by jabatrox** [comment]
Hi @guilhemmarchand 

Ok, I got the the Splunk view as you said (the one from the docs), but the custom fields still do not appear. What I do, is going to the plugin, then:
Get JIRA info > JIRA Service Desk - Get fields description for all projects > I get the 1st screenshot of my previous message.

Then, to get to the 2nd screenshot, I click on Edit > Open in Search, where I see many parsed fields in the left side... but not the custom fields. And actually, if I click on "All fields", I get a screenshot similar to the one in the docs, but without the custom fields:
![image](https://user-images.githubusercontent.com/3520254/93187999-bd0dfe80-f740-11ea-8ea5-fd28ea37a114.png)

I think the reason is that the first project returned from the search is the one in which we don't have custom fields, and for some reason it only parses the fields from that 1st project. As you may see in the 1st screenshot of my previous message, custom fields are present but in the 2nd project listed in the raw search (and the following ones), not in the 1st one (which is not fully expanded in the screenshot)

**Comment by guilhemmarchand** [comment]
@jabatrox 
@hamidbg86 
Implemented in version 1.0.19 which is now live in Splunk Base

A different approach has been taken to finally fix any issue, a new custom command replaced the previous get fields command which had static endpoint path defined.

The new custom command is a generic REST API wrapper for JIRA, which allows you to dynamically define as an argument the endpoint to query, and include any kind of required logic, including JQL queries for any advanced scenario.

Review:
https://ta-jira-service-desk-simple-addon.readthedocs.io/en/latest/userguide.html#jira-rest-api-get-wrapper

The previously existing reports mentioned in this issue remain in place but use the wrapper instead, which can be customised up to any need, or use the custom command to query JIRA depending on your requirements.

Guilhem
